main/CANLogger.c

```diff
@@ -0,0 +1,218 @@
+        rewinddir(d); 
+        while ((dir = readdir(d)) != NULL) {
+            if(strncmp(dir->d_name, fileName, strlen(fileName)) == 0){
+                fileStatus.duplicateNumber++;
+                validName = false;
+                ESP_LOGD(TAG, "%s already exist for base %s\r\n", dir->d_name, fileName);
+                break;
+            }
+        }
+    } while (!validName);
+
+    fileStatus.baseName = &fileName[0];
+    closedir(d);
+    return 0;
+}
+
+/**
+ * Write the csv file header to the first file
+*/
+int writeLogHeader(){
+    if(fileStatus.CANFile == NULL) return -2;
+    if(!xSemaphoreTake(fileStatus.fileMutex, pdMS_TO_TICKS(5))) return -1;
+
+    fprintf(fileStatus.CANFile, "TS,BUS,ID,SEQ,DLC,DATAn\r\n");
+    fileStatus.headerWriten = true;
+
+    xSemaphoreGive(fileStatus.fileMutex);
+    return 0;
+}
+
+/**
+ * Initialize the CAN logger
+*/
+int canLoggerInit(){
+    fileStatus.fileMutex = xSemaphoreCreateMutex();
+    fileStatus.splitNumber = 0;
+    fileStatus.duplicateNumber = 0;
+    fileStatus.filePath = &absPath[0];
+
+    // Make log directory if it doesn't exist
+    struct stat st;
+    if (stat(SD_CARD_CAN_LOG_PATH, &st) == -1) {
+        mkdir(SD_CARD_CAN_LOG_PATH, 0777);
+    }
+
+    if(!createBaseLogName()){
+        canLoggerOpenFile(false);
+        writeLogHeader();
+    }
+
+    return 0;
+}
+
+/**
+ * Process periodic logging tasks
+*/
+int canLoggerUpdate(){
+    if(fileStatus.CANFile == NULL){
+        if(fileStatus.baseName == NULL){
+            createBaseLogName();
+        }
+        canLoggerOpenFile(false);
+    }
+
+    if(!fileStatus.headerWriten)
+        writeLogHeader();
+
+    if(fileStatus.totalBytesWritten > MAX_LOG_SIZE){
+        canLoggerCloseFile();
+        fileStatus.splitNumber++;
+        canLoggerOpenFile(false);
+    }
+
+    if(fileStatus.totalBytesWritten - fileStatus.bytesWrittenAtFlush > FLUSH_SIZE_THRESHOLD || 
+            esp_timer_get_time() - fileStatus.lastFlushTime > FLUSH_LOG_INTERVAL){
+        canLoggerFlushFile();
+    }
+
+    return fileStatus.CANFile != NULL;
+}
+
+/**
+ * Save the received can messages to the log
+ * @param msg the message to log
+*/
+int canLoggerProcessMessage(CANMessage_t const *msg){
+    if(fileStatus.CANFile == NULL) return -2;
+    if(!xSemaphoreTake(fileStatus.fileMutex, pdMS_TO_TICKS(5))) return -1;
+
+    fprintf(fileStatus.CANFile, "%lld,%u,%lx,%lu,%u,",
+        msg->timestamp,
+        msg->bus,
+        msg->ID,
+        msg->SEQ,
+        msg->DLC
+    );
+
+    for(int i=0; i<msg->DLC; i++){
+        fprintf(fileStatus.CANFile,"%x,", msg->payload[i]);
+    }
+
+    fprintf(fileStatus.CANFile,"\r\n");
+
+    fileStatus.totalBytesWritten = ftell(fileStatus.CANFile);
+
+    xSemaphoreGive(fileStatus.fileMutex);
+    return 0;
+}
+
+/**
+ * Open a log file for writing.
+ * @param reopenFile specifies if the file should be reopened to continue writing or a new one should be made
+*/
+int canLoggerOpenFile(bool reopenFile){
```

Comment:
This is a little clearer in my opinion
```suggestion
int canLoggerOpenFile(bool append){
```